### PR TITLE
 Fixes for issues #1333 and #1381 

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
This completes the implementation of handling named arguments.

### CODE REVIEW

File: frontends/p4/defaultArguments.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    const IR::Node* postorder(IR::MethodCallExpression* expression) override;
+    const IR::Node* postorder(IR::Declaration_Instance* inst) override;
+    const IR::Node* postorder(IR::ConstructorCallExpression* ccc) override;
+};
+
+class DefaultArguments : public PassManager {
+ public:
+    DefaultArguments(ReferenceMap* refMap, TypeMap* typeMap) {
+        setName("DefaultArguments");
+        passes.push_back(new TypeChecking(refMap, typeMap));
```
Comment (cc10512) [inline frontends/p4/defaultArguments.h]: Type checking first and then adding the default arguments? I would have thought you would check after you've added them.

File: frontends/p4/typeChecking/typeChecker.cpp [hunk 2]
```diff
@@ -1406,6 +1407,14 @@ const IR::Node* TypeInference::postorder(IR::Parameter* param) {
         return param;
     BUG_CHECK(!paramType->is<IR::Type_Type>(), "%1%: unexpected type", paramType);
 
+    if (param->defaultValue != nullptr) {
+        auto init = assignment(param, paramType, param->defaultValue);
+        if (init == nullptr)
+            return param;
+        if (param->defaultValue != init)
+            param->defaultValue = init;
```
Comment (cc10512) [inline frontends/p4/typeChecking/typeChecker.cpp]: Are you overloading `defaultValue`? Or is assignment likely to return a simplified expression?

File: frontends/p4/typeChecking/typeUnification.cpp [hunk 3]
```diff
@@ -123,9 +123,9 @@ bool TypeUnification::unifyCall(const IR::Node* errorPosition,
             ++paramIt;
     }
 
-    // Check remaining parameters: they must be all optional
+    // Check remaining parameters: they must be all optional or have default values
     for (auto p : left) {
-        bool opt = p.second->isOptional();
+        bool opt = p.second->isOptional() || p.second->defaultValue != nullptr;
```
Comment (cc10512) [inline frontends/p4/typeChecking/typeUnification.cpp]: +1. And soon we can get rid of @optional!

File: frontends/p4/defaultArguments.h (hunk 1 above)
Comment (mbudiu-vmw) [reply frontends/p4/defaultArguments.h]: I tried to to it the other way round, but it's much harder.
The problem is that given an expression a.b.c.isValid() to know what the function isValid is you need to know more about a.b.c and that information comes from type checking.

File: frontends/p4/typeChecking/typeChecker.cpp (hunk 2 above)
Comment (mbudiu-vmw) [reply frontends/p4/typeChecking/typeChecker.cpp]: Assignment does type inference, and may return a new expression with modified types. In particular, `int` types may be converted to `bit<>` types.

File: frontends/p4/typeChecking/typeUnification.cpp (hunk 3 above)
Comment (mbudiu-vmw) [reply frontends/p4/typeChecking/typeUnification.cpp]: At this point we don't have anything to replace optional.

**Pull request merged**